File: typescript/packages/postgres/src/placeholders.ts
// The portable subset's `?` placeholders in pg's form, $1..$n, rewritten outside quoted literals
// and identifiers (spec/conformance/vectors/sql-placeholders.json). A literal `%` is left alone:
// only psycopg needs it doubled.

const cache = new Map<string, string>();

export function numbered(sql: string): string {
  const found = cache.get(sql);
  if (found !== undefined) return found;
  let out = "";
  let quote: string | undefined;
  let n = 0;
  for (const char of sql) {
    if (quote !== undefined) {
      if (char === quote) quote = undefined;
      out += char;
    } else if (char === "'" || char === '"') {
      quote = char;
      out += char;
    } else if (char === "?") {
      n += 1;
      out += `$${n}`;
    } else out += char;
  }
  cache.set(sql, out);
  return out;
}

File: python/src/threads/postgres/placeholders.py
"""The portable subset's `?` placeholders in psycopg's form: `%s`, with every literal `%`
doubled, outside quoted literals and identifiers (spec/conformance/vectors/sql-placeholders.json).
"""

from functools import cache


@cache
def psycopg(sql: str) -> str:
    out: list[str] = []
    quote: str | None = None
    for char in sql:
        if quote is not None:
            if char == quote:
                quote = None
            out.append("%%" if char == "%" else char)
        elif char in "'\"":
            quote = char
            out.append(char)
        elif char == "?":
            out.append("%s")
        else:
            out.append("%%" if char == "%" else char)
    return "".join(out)
